### PR TITLE
Add skeleton-related traits 

### PR DESCRIPTION
**Pull request by chengshuyi**
Add three traits, SkelBuilder, OpenSkel, and Skel, which are mainly used in the scenario of managing multiple skeletons, so that multiple skeletons can have the same abstract type.

Best regards and thank you in advance.

### CODE REVIEW

**Comment by chengshuyi** [comment]
#442

File: libbpf-rs/src/skeleton.rs [hunk 1]
```diff
@@ -313,3 +316,26 @@ impl<'a> Drop for ObjectSkeletonConfig<'a> {
...
+pub trait OpenSkel {
+    type Output: Skel;
+    fn load(self) -> Result<Self::Output>;
+    fn open_object(&self) -> &OpenObject;
+    fn open_object_mut(&mut self) -> &mut OpenObject;
+}
+
+pub trait Skel {
+    fn attach(&mut self) -> Result<()> {
+        unimplemented!()
```
Comment (danielocfb) [inline libbpf-rs/src/skeleton.rs:337]: Why do we need this default definition?

File: libbpf-rs/src/skeleton.rs [hunk 2]
```diff
@@ -313,3 +316,26 @@ impl<'a> Drop for ObjectSkeletonConfig<'a> {
         unsafe { Box::from_raw(self.inner.obj) };
     }
 }
+
+pub trait SkelBuilder<'a> {
```
Comment (danielocfb) [inline libbpf-rs/src/skeleton.rs:320]: Can we please add documentation for these traits and their methods?

**Review by danielocfb — CHANGES_REQUESTED** [review]
Thanks! Changes seem fine for the most part, but I think we should make sure that those traits show up in the documentation. Currently the entire skeleton module is hidden from it. So we need to rework that. E.g.:
```patch
diff --git examples/capable/src/main.rs examples/capable/src/main.rs
index 89f211..e68ad5 100644
--- examples/capable/src/main.rs
+++ examples/capable/src/main.rs
@@ -9,9 +9,9 @@ use std::str::FromStr;
 use anyhow::bail;
 use anyhow::Result;
 use clap::Parser;
-use libbpf_rs::skeleton::OpenSkel;
-use libbpf_rs::skeleton::Skel;
-use libbpf_rs::skeleton::SkelBuilder;
+use libbpf_rs::skel::OpenSkel;
+use libbpf_rs::skel::Skel;
+use libbpf_rs::skel::SkelBuilder;
 use libbpf_rs::PerfBufferBuilder;
 use phf::phf_map;
 use plain::Plain;
diff --git examples/runqslower/src/main.rs examples/runqslower/src/main.rs
index 5cede7..1ab237 100644
--- examples/runqslower/src/main.rs
+++ examples/runqslower/src/main.rs
@@ -5,9 +5,9 @@ use core::time::Duration;
 use anyhow::bail;
 use anyhow::Result;
 use clap::Parser;
-use libbpf_rs::skeleton::OpenSkel;
-use libbpf_rs::skeleton::Skel;
-use libbpf_rs::skeleton::SkelBuilder;
+use libbpf_rs::skel::OpenSkel;
+use libbpf_rs::skel::Skel;
+use libbpf_rs::skel::SkelBuilder;
 use libbpf_rs::PerfBufferBuilder;
 use plain::Plain;
 use time::macros::format_description;
diff --git examples/tc_port_whitelist/src/main.rs examples/tc_port_whitelist/src/main.rs
index 639e44..e4d514 100644
--- examples/tc_port_whitelist/src/main.rs
+++ examples/tc_port_whitelist/src/main.rs
@@ -1,8 +1,8 @@
 use anyhow::bail;
 use anyhow::Result;
 use clap::Parser;
-use libbpf_rs::skeleton::OpenSkel;
-use libbpf_rs::skeleton::SkelBuilder;
+use libbpf_rs::skel::OpenSkel;
+use libbpf_rs::skel::SkelBuilder;
 use libbpf_rs::MapFlags;
 use libbpf_rs::TcHookBuilder;
 use libbpf_rs::TC_CUSTOM;
diff --git examples/tproxy/src/main.rs examples/tproxy/src/main.rs
index 8113fd..ff802b 100644
--- examples/tproxy/src/main.rs
+++ examples/tproxy/src/main.rs
@@ -9,8 +9,8 @@ use std::time::Duration;
 use anyhow::Context;
 use anyhow::Result;
 use clap::Parser;
-use libbpf_rs::skeleton::OpenSkel;
-use libbpf_rs::skeleton::SkelBuilder;
+use libbpf_rs::skel::OpenSkel;
+use libbpf_rs::skel::SkelBuilder;
 use libbpf_rs::TcHookBuilder;
 use libbpf_rs::TC_INGRESS;

diff --git libbpf-cargo/src/gen/mod.rs libbpf-cargo/src/gen/mod.rs
index 45c722..585398 100644
--- libbpf-cargo/src/gen/mod.rs
+++ libbpf-cargo/src/gen/mod.rs
@@ -214,9 +214,9 @@ fn gen_skel_c_skel_constructor(skel: &mut String, object: &mut BpfObj, name: &st
     write!(
         skel,
         r#"
-        fn build_skel_config() -> libbpf_rs::Result<libbpf_rs::skeleton::ObjectSkeletonConfig<'static>>
+        fn build_skel_config() -> libbpf_rs::Result<libbpf_rs::__internal_skel::ObjectSkeletonConfig<'static>>
         {{
-            let mut builder = libbpf_rs::skeleton::ObjectSkeletonConfigBuilder::new(DATA);
+            let mut builder = libbpf_rs::__internal_skel::ObjectSkeletonConfigBuilder::new(DATA);
             builder
                 .name("{name}")
         "#,
@@ -663,7 +663,7 @@ fn gen_skel_contents(_debug: bool, raw_obj_name: &str, obj_file_path: &Path) ->
            #[allow(clippy::upper_case_acronyms)]
            mod imp {{
            use libbpf_rs::libbpf_sys;
-           use libbpf_rs::skeleton::{{SkelBuilder, OpenSkel, Skel}};
+           use libbpf_rs::skel::{{SkelBuilder, OpenSkel, Skel}};
         "#
     )?;

@@ -747,7 +747,7 @@ fn gen_skel_contents(_debug: bool, raw_obj_name: &str, obj_file_path: &Path) ->
         r#"
         pub struct Open{name}Skel<'a> {{
             pub obj: libbpf_rs::OpenObject,
-            skel_config: libbpf_rs::skeleton::ObjectSkeletonConfig<'a>,
+            skel_config: libbpf_rs::__internal_skel::ObjectSkeletonConfig<'a>,
         }}

         impl<'a> OpenSkel for Open{name}Skel<'a> {{
@@ -803,7 +803,7 @@ fn gen_skel_contents(_debug: bool, raw_obj_name: &str, obj_file_path: &Path) ->
         r#"
         pub struct {name}Skel<'a> {{
             pub obj: libbpf_rs::Object,
-            skel_config: libbpf_rs::skeleton::ObjectSkeletonConfig<'a>,
+            skel_config: libbpf_rs::__internal_skel::ObjectSkeletonConfig<'a>,
         "#,
         name = &obj_name,
     )?;
diff --git libbpf-rs/src/lib.rs libbpf-rs/src/lib.rs
index f8684c..8e390a 100644
--- libbpf-rs/src/lib.rs
+++ libbpf-rs/src/lib.rs
@@ -87,12 +87,23 @@ mod print;
 mod program;
 pub mod query;
 mod ringbuf;
-/// Used for skeleton -- an end user may not consider this API stable
-#[doc(hidden)]
-pub mod skeleton;
+mod skeleton;
 mod tc;
 mod util;

+/// Used for skeleton -- an end user may not consider this API stable
+#[doc(hidden)]
+pub mod __internal_skel {
+    pub use super::skeleton::*;
+}
+
+/// Skeleton related definitions.
+pub mod skel {
+    pub use super::skeleton::OpenSkel;
+    pub use super::skeleton::Skel;
+    pub use super::skeleton::SkelBuilder;
+}
+
 pub use libbpf_sys;

 pub use crate::btf::Btf;
```

**Comment by chengshuyi** [comment]
> 

Ok, will do.

File: libbpf-rs/src/skeleton.rs (hunk 1 above)
Comment (chengshuyi) [reply libbpf-rs/src/skeleton.rs:337]: When there is no eBPF program in bpf.c, libbpf-cargo will not generate the attach function, which will cause an error in compiling the test_skeleton_empty_source test, so the default implementation is provided here. The following code is a snippet of the attach function generated by libbpf-cargo
```rust
fn gen_skel_attach(skel: &mut String, object: &mut BpfObj, obj_name: &str) -> Result<()> {
    if ProgIter::new(object.as_mut_ptr()).next().is_none() {
        return Ok(());
    }

    write!(
    ...
```

File: libbpf-rs/src/skeleton.rs (hunk 2 above)
Comment (chengshuyi) [reply libbpf-rs/src/skeleton.rs:320]: Ok, will do

File: libbpf-rs/src/skeleton.rs (hunk 1 above)
Comment (danielocfb) [reply libbpf-rs/src/skeleton.rs:337]: Interesting. I haven't looked into it in detail, but intuitively I'd say we should probably change that and generate the function unconditionally. But I guess we can do that as a follow-up, if you don't want to do it right now.

**Comment by chengshuyi** [comment]
> E: Failed to fetch http://azure.archive.ubuntu.com/ubuntu/pool/main/e/elfutils/libelf-dev_0.186-1build1_amd64.deb  Could not connect to azure.archive.ubuntu.com:80 (52.147.219.192), connection timed out  Could not connect to azure.archive.ubuntu.com:80 (52.147.219.192), connection timed out
Hi, @danielocfb  it seems ubuntu package server is not stable which causes ci failing.

**Comment by danielocfb** [comment]
> > E: Failed to fetch http://azure.archive.ubuntu.com/ubuntu/pool/main/e/elfutils/libelf-dev_0.186-1build1_amd64.deb  Could not connect to azure.archive.ubuntu.com:80 (52.147.219.192), connection timed out  Could not connect to azure.archive.ubuntu.com:80 (52.147.219.192), connection timed out
> 
> Hi, @danielocfb it seems ubuntu package server is not stable which causes ci failing.

Sigh. I've restarted the failed jobs and they passed.

**Review by danielocfb — APPROVED** [review]
Looks good to me.

**Comment by danielocfb** [comment]
Merged with minor adjustments. Thanks!

**Pull request closed without merging**